File: pjecalc-fonte/negocio/br/jus/trt8/pjecalc/negocio/dominio/calculo/atualizacao/regras/DataGeralParametrosAtualizacaoValidRule.java
/*
 * Decompiled with CFR 0.152.
 */
package br.jus.trt8.pjecalc.negocio.dominio.calculo.atualizacao.regras;

import br.jus.trt8.pjecalc.base.comum.HelperDate;
import br.jus.trt8.pjecalc.base.comum.Utils;
import br.jus.trt8.pjecalc.base.comum.validadores.ValidatorContext;
import br.jus.trt8.pjecalc.negocio.comum.validators.ValidRule;
import br.jus.trt8.pjecalc.negocio.comum.validators.ValidValueValidator;
import br.jus.trt8.pjecalc.negocio.constantes.Mensagens;
import br.jus.trt8.pjecalc.negocio.dominio.calculo.atualizacao.CombinacaoDeIndice;
import br.jus.trt8.pjecalc.negocio.dominio.calculo.atualizacao.CombinacaoDeJuros;
import br.jus.trt8.pjecalc.negocio.dominio.calculo.atualizacao.ParametrosDeAtualizacao;
import java.util.Date;

public class DataGeralParametrosAtualizacaoValidRule
implements ValidRule {
    private static final String DATA_HOJE = "dataHoje";
    private static final String DATA_ADMISSAO = "calculo.dataAdmissao";
    private static final String DATA_ULTIMA_ATUALIZACAO = "calculoExterno.dataUltimaAtualizacao";
    private static final String DATA_AJUIZAMENTO = "calculo.dataAjuizamento";
    private int flag = 0;

    @Override
    public Mensagens getMessage() {
        if (this.flag == 0) {
            return Mensagens.MSG0007;
        }
        return Mensagens.MSG0010;
    }

    @Override
    public boolean isValid(ValidValueValidator validator, ValidatorContext context) {
        Date admissao;
        ParametrosDeAtualizacao parametrosDeAtualizacao = null;
        boolean isCombinacaoJuros = false;
        if (context.getBean() instanceof ParametrosDeAtualizacao) {
            parametrosDeAtualizacao = (ParametrosDeAtualizacao)context.getBean();
        } else if (context.getBean() instanceof CombinacaoDeIndice) {
            parametrosDeAtualizacao = ((CombinacaoDeIndice)context.getBean()).getParametrosDeAtualizacao();
        } else if (context.getBean() instanceof CombinacaoDeJuros) {
            parametrosDeAtualizacao = ((CombinacaoDeJuros)context.getBean()).getParametrosDeAtualizacao();
            isCombinacaoJuros = true;
        } else {
            return true;
        }
        boolean isCalculoExterno = parametrosDeAtualizacao.getCalculo().isCalculoExterno();
        Date aPartirDe = (Date)context.getValue();
        if (isCombinacaoJuros) {
            Date dataLimite;
            Date date = isCalculoExterno ? parametrosDeAtualizacao.getCalculo().getDataDeLiquidacao() : (dataLimite = parametrosDeAtualizacao.getAplicarJurosFasePreJudicial() != false ? parametrosDeAtualizacao.getCalculo().getDataAdmissao() : parametrosDeAtualizacao.getCalculo().getDataAjuizamento());
            if (Utils.nulo(aPartirDe) || Utils.nulo(dataLimite) || HelperDate.dateAfter(aPartirDe, dataLimite)) {
                return true;
            }
            validator.getParameters().put("1", isCalculoExterno ? DATA_ULTIMA_ATUALIZACAO : (parametrosDeAtualizacao.getAplicarJurosFasePreJudicial() != false ? DATA_ADMISSAO : DATA_AJUIZAMENTO));
            return false;
        }
        Date date = admissao = isCalculoExterno ? parametrosDeAtualizacao.getCalculo().getDataDeLiquidacao() : parametrosDeAtualizacao.getCalculo().getDataAdmissao();
        if (Utils.naoNulo(aPartirDe) && Utils.naoNulo(admissao)) {
            HelperDate hoje = HelperDate.getInstance();
            hoje.removeTime();
            if (HelperDate.dateBeforeOrEquals(aPartirDe, admissao)) {
                validator.getParameters().put("1", isCalculoExterno ? DATA_ULTIMA_ATUALIZACAO : DATA_ADMISSAO);
                return false;
            }
            if (hoje.lessThen(aPartirDe)) {
                this.flag = 1;
                validator.getParameters().put("1", DATA_HOJE);
                return false;
            }
        }
        return true;
    }
}

File: src/lib/pjecalc/core/dominio/calculo/atualizacao/regras/data-geral-parametros-atualizacao-valid-rule.ts
/**
 * Porte 1:1 de DataGeralParametrosAtualizacaoValidRule.java (76 linhas).
 *
 * Regra de validação geral para datas de ParametrosDeAtualizacao,
 * CombinacaoDeIndice e CombinacaoDeJuros.
 *
 * Verifica que a "data a partir de" é:
 *   - Posterior à data de admissão (ou data de liquidação, em cálculo externo).
 *   - Não posterior a hoje.
 *   - Para CombinacaoDeJuros: também precisa ser posterior à data de
 *     ajuizamento/admissão conforme flag aplicarJurosFasePreJudicial.
 *
 * Ref: pjecalc-fonte/.../dominio/calculo/atualizacao/regras/DataGeralParametrosAtualizacaoValidRule.java
 */
import { HelperDate } from '../../../../base/comum/helper-date';
import { CombinacaoDeIndice } from '../combinacao-de-indice';
import { CombinacaoDeJuros } from '../combinacao-de-juros';
import { ParametrosDeAtualizacao } from '../parametros-de-atualizacao';

export interface DataGeralValidResult {
  valid: boolean;
  message?: string;
  detailParam?: string;
}

export class DataGeralParametrosAtualizacaoValidRule {
  /**
   * Valida a data "aPartirDe" no bean.
   * bean pode ser ParametrosDeAtualizacao, CombinacaoDeIndice ou CombinacaoDeJuros.
   */
  static isValid(
    aPartirDe: Date | null,
    bean: ParametrosDeAtualizacao | CombinacaoDeIndice | CombinacaoDeJuros,
  ): DataGeralValidResult {
    let parametros: ParametrosDeAtualizacao | null;
    let isCombinacaoJuros = false;

    if (bean instanceof ParametrosDeAtualizacao) {
      parametros = bean;
    } else if (bean instanceof CombinacaoDeIndice) {
      parametros = bean.getParametrosDeAtualizacao();
    } else if (bean instanceof CombinacaoDeJuros) {
      parametros = bean.getParametrosDeAtualizacao();
      isCombinacaoJuros = true;
    } else {
      return { valid: true };
    }

    if (!parametros) return { valid: true };
    const calc = parametros.getCalculo() as unknown as {
      isCalculoExterno?: () => boolean;
      getDataDeLiquidacao?: () => Date | null;
      getDataAdmissao?: () => Date | null;
      getDataAjuizamento?: () => Date | null;
    } | null;
    if (!calc) return { valid: true };
    const isCalculoExterno = calc.isCalculoExterno?.() ?? false;

    if (isCombinacaoJuros) {
      const dataLimite = isCalculoExterno
        ? calc.getDataDeLiquidacao?.()
        : (parametros.getAplicarJurosFasePreJudicial()
          ? calc.getDataAdmissao?.()
          : calc.getDataAjuizamento?.());
      if (!aPartirDe || !dataLimite || HelperDate.dateAfter(aPartirDe, dataLimite)) {
        return { valid: true };
      }
      return {
        valid: false,
        message: 'MSG0007',
        detailParam: isCalculoExterno
          ? 'calculoExterno.dataUltimaAtualizacao'
          : (parametros.getAplicarJurosFasePreJudicial() ? 'calculo.dataAdmissao' : 'calculo.dataAjuizamento'),
      };
    }

    const admissao = isCalculoExterno
      ? calc.getDataDeLiquidacao?.()
      : calc.getDataAdmissao?.();

    if (aPartirDe && admissao) {
      const hoje = new Date();
      hoje.setHours(0, 0, 0, 0);
      if (HelperDate.dateBeforeOrEquals(aPartirDe, admissao)) {
        return {
          valid: false,
          message: 'MSG0007',
          detailParam: isCalculoExterno ? 'calculoExterno.dataUltimaAtualizacao' : 'calculo.dataAdmissao',
        };
      }
      if (HelperDate.dateBefore(hoje, aPartirDe)) {
        return { valid: false, message: 'MSG0010', detailParam: 'dataHoje' };
      }
    }
    return { valid: true };
  }
}
